### PR TITLE
8319705: RISC-V: signumF/D intrinsics fails compiler/intrinsics/math/TestSignumIntrinsic.java

### PR DESCRIPTION
**Pull request by Hamlin-Li**
Hi,
Can you review the patch to fix the bug introduced by [JDK-8317971](https://bugs.openjdk.org/browse/JDK-8317971) which implemented signumF/D intrinsics.
But seems the intrinsics failed since day 1 (I checkouted the commit to verify it.)
Thanks

## Test
hotspot tests:
```
test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java
test/hotspot/jtreg/compiler/c2/PolynomialRoot.java
```

other jdk tests contains Math.signum(...):
```
test/jdk/java/math/BigInteger/BigIntegerTest.java 
test/jdk/java/math/BigInteger/largeMemory/DivisionOverflow.java
test/jdk/java/math/BigInteger/SerializationTests.java
test/jdk/java/math/BigInteger/ModInvTime.java
test/jdk/java/math/BigDecimal/SquareRootTests.java
test/jdk/java/math/BigDecimal/DivideTests.java
test/jdk/java/lang/Math/IeeeRecommendedTests.java
test/jdk/java/lang/Math/IeeeRemainderTests.java
```

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8319705](https://bugs.openjdk.org/browse/JDK-8319705): RISC-V: signumF/D intrinsics fails compiler/intrinsics/math/TestSignumIntrinsic.java (**Bug** - P2)

### Reviewers
 * [Fei Yang](https://openjdk.org/census#fyang) (@RealFYang - **Reviewer**) ⚠️ Review applies to [85de9252](https://git.openjdk.org/jdk/pull/16557/files/85de9252b8ad9ab5f44608a681c7670c7eafe6e5)
 * [Vladimir Kempik](https://openjdk.org/census#vkempik) (@VladimirKempik - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16557/head:pull/16557` \
`$ git checkout pull/16557`

Update a local copy of the PR: \
`$ git checkout pull/16557` \
`$ git pull https://git.openjdk.org/jdk.git pull/16557/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16557`

View PR using the GUI difftool: \
`$ git pr show -t 16557`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16557.diff">https://git.openjdk.org/jdk/pull/16557.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16557#issuecomment-1801767202)

### CODE REVIEW

**Comment by RealFYang** [comment]
Do you have more details about the problem? Those tests are passing on hifive unmatched board using today's jdk master.

**Comment by Hamlin-Li** [comment]
I ran it via qemu, triggered the test with Xcomp to make sure the intrinsic is installed.
It just failed with fastdebug version, no special config.

**Comment by VladimirKempik** [comment]
So maybe the test compiler/intrinsics/math/TestSignumIntrinsic.java doesn't test anything on risc-v in default mode ?

**Comment by VladimirKempik** [comment]
it's not clear what's wrong with original version
instead of using t0 as temp register inside signumFD, you explicetly requested tmp register from C2

but t0 is not allocatable ( by c2 reg alloc) register and often used freely.

I don't get it what failed and why it doesn't fail anymore

**Comment by RealFYang** [comment]
OK. I can now reproduce the bug with following command on sifive unmatched board:
```
make test TEST="./test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java" JTREG="VM_OPTIONS=-XX:-TieredCompilation"
```
Error message:
```
java.lang.RuntimeException: Unexpected double result from 123.4: expected 1.0 to equal 123.4
        at jdk.test.lib.Asserts.fail(Asserts.java:634)
        at jdk.test.lib.Asserts.assertEquals(Asserts.java:205)
        at jdk.test.lib.Asserts.assertEQ(Asserts.java:178)
        at compiler.intrinsics.math.TestSignumIntrinsic.doubleTest(TestSignumIntrinsic.java:114)
        at compiler.intrinsics.math.TestSignumIntrinsic.main(TestSignumIntrinsic.java:91)
        at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
        at java.base/java.lang.reflect.Method.invoke(Method.java:580)
        at com.sun.javatest.regtest.agent.MainWrapper$MainTask.run(MainWrapper.java:138)
        at java.base/java.lang.Thread.run(Thread.java:1570)

JavaTest Message: Test threw exception: java.lang.RuntimeException: Unexpected double result from 123.4: expected 1.0 to equal 123.4
```

But it does not always reproduce as I think it depends on C2 register allocation.
For `C2_MacroAssembler::signum_fp`, it has three FP registers: `dst`, `src`, and `one`. And it is possible that `dst` and `one` could be assigned the same FP register. This will be a problem after change [1]. This change copies `src` to `dst` on entry, thus modifies `one` in that case. This is erroneous as `one` is still alive here and will be used later by `fsgnj_d/s` if `src` is not NaN or +/-0.0.

[1] https://github.com/openjdk/jdk/pull/16186/commits/867d6e8ea2a53ae71fb8fd341892e8f88347ec07

File: src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp [hunk 1]
```diff
@@ -1658,25 +1658,20 @@ void C2_MacroAssembler::round_double_mode(FloatRegister dst, FloatRegister src,
 // otherwise return +/- 1.0 using sign of input.
 // one - gives us a floating-point 1.0 (got from matching rule)
 // bool is_double - specifies single or double precision operations will be used.
-void C2_MacroAssembler::signum_fp(FloatRegister dst, FloatRegister src, FloatRegister one, bool is_double) {
-  Register tmp1 = t0;
-
+void C2_MacroAssembler::signum_fp(FloatRegister dst, FloatRegister one, Register tmp, bool is_double) {
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp:1661]: I think we could still make use of scratch register `t0` in this function. That will save us one GP register.

File: src/hotspot/cpu/riscv/riscv.ad [hunk 2]
```diff
@@ -7601,22 +7601,24 @@ instruct copySignF_reg(fRegF dst, fRegF src1, fRegF src2) %{
   ins_pipe(fp_dop_reg_reg_s);
 %}
 
-instruct signumD_reg(fRegD dst, fRegD src, immD zero, fRegD one) %{
-  match(Set dst (SignumD src (Binary zero one)));
-  format %{ "signumD  $dst, $src" %}
+instruct signumD_reg(fRegD dst, immD zero, fRegD one, iRegLNoSp tmp) %{
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv.ad:7604]: No need to reserve `tmp` here. See my previous comment.

File: src/hotspot/cpu/riscv/riscv.ad [hunk 3]
```diff
@@ -7601,22 +7601,24 @@ instruct copySignF_reg(fRegF dst, fRegF src1, fRegF src2) %{
...
+    __ signum_fp(as_FloatRegister($dst$$reg), as_FloatRegister($one$$reg),
+                 as_Register($tmp$$reg), true /* is_double */);
   %}
   ins_pipe(pipe_class_default);
 %}
 
-instruct signumF_reg(fRegF dst, fRegF src, immF zero, fRegF one) %{
-  match(Set dst (SignumF src (Binary zero one)));
-  format %{ "signumF  $dst, $src" %}
+instruct signumF_reg(fRegF dst, immF zero, fRegF one, iRegLNoSp tmp) %{
```
Comment (RealFYang) [inline src/hotspot/cpu/riscv/riscv.ad:7615]: No need to reserve `tmp` here. See my previous comment.

**Review by RealFYang — CHANGES_REQUESTED** [review]

**Comment by VladimirKempik** [comment]
> it's not clear what's wrong with original version instead of using t0 as temp register inside signumFD, you explicetly requested tmp register from C2
> 
> but t0 is not allocatable ( by c2 reg alloc) register and often used freely.
> 
> I don't get it what failed and why it doesn't fail anymore

I see now, when dst register is the same as register one, we've got the issue, and we can provoke such behaviour from RegAlloc only when disabling C1. test needs to be modified as well

File: src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp (hunk 1 above)
Comment (Hamlin-Li) [reply src/hotspot/cpu/riscv/c2_MacroAssembler_riscv.cpp]: Sure, it make sense.

**Comment by Hamlin-Li** [comment]
@RealFYang Thank you for exploring and sharing the information.

**Review by RealFYang — APPROVED** [review]

**Comment by VladimirKempik** [comment]
Please don' integrate until testcase update info is clear, we are working on it

**Comment by Ilyagavrilin** [comment]
Hi, could you, please,  add this small update to jtreg: hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java:
```
diff --git a/test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java b/test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java
index 65d261bb8e2..a5546ba9e5c 100644
--- a/test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java
+++ b/test/hotspot/jtreg/compiler/intrinsics/math/TestSignumIntrinsic.java
@@ -35,6 +35,10 @@
  *      -XX:+IgnoreUnrecognizedVMOptions -XX:+UnlockDiagnosticVMOptions
  *      -XX:-UseSignumIntrinsic -XX:+UseCopySignIntrinsic
  *      compiler.intrinsics.math.TestSignumIntrinsic
+ * @run main/othervm
+ *      -XX:+IgnoreUnrecognizedVMOptions -XX:+UnlockDiagnosticVMOptions
+ *      -Xcomp -XX:+UseSignumIntrinsic
+ *      compiler.intrinsics.math.TestSignumIntrinsic
  */
 
 package compiler.intrinsics.math;
```

**Review by VladimirKempik — APPROVED** [review]

**Comment by Hamlin-Li** [comment]
> @Hamlin-Li do you have GHA enabled for your repo ? It would be nice to let tier1 in gha pass to check updated test on all x86 platforms

Yes, I think so.

**Comment by Hamlin-Li** [comment]
/integrate

Thanks @RealFYang @VladimirKempik @Ilyagavrilin for your review.

**Pull request closed without merging**